include/nbl/asset/utils/CSPIRVIntrospector.h

```diff
@@ -13,172 +17,657 @@
-					double f64;
-					float f32;
-				} defaultValue;
-			};
-			//! Sorted by `id`
-			core::vector<SSpecConstant> specConstants;
-			//! Each vector is sorted by `binding`
-			core::vector<SShaderResourceVariant> descriptorSetBindings[4];
-			//! Sorted by `location`
-			core::vector<SShaderInfoVariant> inputOutput;
-
-			//! Push constants uniform block
-			struct {
-				bool present;
-				core::string name;
-				SShaderPushConstant info;
-			} pushConstant;
-
-			bool canSpecializationlesslyCreateDescSetFrom() const
-			{
-				for (const auto& descSet : descriptorSetBindings)
+				//! Descriptors
+				template<bool Mutable=false>
+				struct SDescriptorVarInfo final : SDescriptorInfo
+				{
+					struct SRWDescriptor
+					{
+						uint8_t readonly : 1 = false;
+						uint8_t writeonly : 1 = false;
+					};
+					struct SImage
+					{
+						IImageView<ICPUImage>::E_TYPE viewType : 3 = IImageView<ICPUImage>::E_TYPE::ET_2D;
+						uint8_t shadow : 1 = false;
+					};
+					struct SCombinedImageSampler final : SImage
+					{
+						uint8_t multisample : 1;
+					};
+					struct SStorageImage final : SRWDescriptor, SImage
+					{
+						// `EF_UNKNOWN` means that Shader will use the StoreWithoutFormat or LoadWithoutFormat capability
+						E_FORMAT format = EF_UNKNOWN;
+					};
+					struct SUniformTexelBuffer final
+					{
+					};
+					struct SStorageTexelBuffer final : SRWDescriptor
+					{
+					};
+					struct SUniformBuffer final : SMemoryBlock<Mutable>
+					{
+						size_t size = 0;
+					};
+					struct SStorageBuffer final : SRWDescriptor, SMemoryBlock<Mutable>
+					{
+						template<bool C=!Mutable>
+						inline std::enable_if_t<C,bool> isLastMemberRuntimeSized() const
+						{
+							if (type->memberCount)
+								return type->memberTypes()[type->memberCount-1].count.front().isRuntimeSized();
+							return false;
+						}
+						template<bool C=!Mutable>
+						inline std::enable_if_t<C,size_t> getRuntimeSize(const size_t lastMemberElementCount) const
+						{
+							if (isLastMemberRuntimeSized())
+							{
+								const auto& lastMember = type->memberTypes()[type->memberCount-1];
+								assert(!lastMember.count.front().isSpecConstantID);
+								return sizeWithoutLastMember+lastMemberElementCount*type->memberStrides()[type->memberCount-1];
+							}
+							return sizeWithoutLastMember;
+						}
+
+						//! Use `getRuntimeSize` for size of the struct with assumption of passed number of elements.
+						//! Need special handling if last member is rutime-sized array (e.g. buffer SSBO `buffer { float buf[]; }`)
+						size_t sizeWithoutLastMember;
+					};
+					struct SInputAttachment final
+					{
+						uint32_t index;
+					};
+
+					inline bool isArray() const {return !count.empty();}
+					inline bool isRunTimeSized() const {return isArray() ? count[0].value == 0 : false;}
+
+					//! Note: for SSBOs and UBOs it's the block name
+					span_t<char,Mutable> name = {};
+					span_t<SArrayInfo,Mutable> count = {};
```

Comment:
maybe we'll need to use different type than `SArrayInfo` if there's an ifff relationship between `OpRuntimeArray` and `VARIABLE_DESCRIPTOR_COUNT_BIT`
https://discord.com/channels/427551838099996672/427951526967902218/1230618569931227146

Reply:
I read the spec https://registry.khronos.org/vulkan/specs/1.3/html/vkspec.html#interfaces-resources-setandbinding a bit and found:

> The array must be a single-dimensional array and size of the array must be no larger than the number of descriptors in the binding.

Therefore we shouldn't use `span_t<SArrayInfo>` and just a single `SArrayInfo`

It would be good to leave this spec quote here and refer to section 15.5.3

Reply:
We can still use `SArrayInfo` as the stage introspection doesn't need to worry about the case of a Vulkan Layout SBinding having a runtime flag while also having a SPIR-V module that defines the binding as a static sized array